flink-core/src/main/java/org/apache/flink/core/execution/JobListener.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.execution;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.JobID;
+
+/**
+ * A listener that is notified on specific job status changed.
+ */
+@PublicEvolving
+public interface JobListener {
+
+	/**
+	 * Callback on job submission succeeded.
+	 *
+	 * @param jobID of the submitted job
+	 */
+	void onJobSubmitted(JobID jobID);
```

Comment:
2 comments:
1. It would be better to add `onJobExecuted` which is called when job is finished. Otherwise Zeppelin don't know when job is finished and have to always put the data of job in memory.
2. It might be better to pass `JobClient` instead of `JobID` to `onJobSubmitted`. That means after job is submitted, user can populate the job via this `JobClient`.

Reply:
1. OK I'll try to add `onJobExecuted` but it should be only works when user call `execute`(instead of `executeAsync`)

2. Please see my comment above where I list overhead for doing so. It is a design decision but let's make sure the risk is delivered.

Reply:
2. Or in another way my concern can be addressed. Would you like to handle a possibly closed `JobClient` `onJobSubmitted`? If the job finished, you cannot do too much thing. And a call like `getJobStatus` will failed since `JobClient` got closed. It should be handled by user code. If it is expected, I'm ok with passing `JobClient`.

Reply:
1. It is fine for `onJobExecuted` in `execute`.
2. I think user need to take this risk as long as he wants to control job lifecycle via `JobClient`. Since `JobClient` is a user-facing API, the risk is always there no matter we use it here or somewhere else.

Reply:
For `onJobExecuted`, what do you want to be called back? Is it necessary to be a `JobExecutionResult` i.e. execute attach? Or a `DetachedJobExecutionResult` i.e. execute detached is ok?

Reply:
@zjffdu thanks for your reply. I will then pass `JobClient` and try to document the attention.